### DISCUSSION TITLE
Extending notify.py to draw unmanaged windows for notifications

### DISCUSSION THREAD

**Discussion by m-col** [qtile/qtile#2772] [Ideas]
[notify.py](https://github.com/qtile/qtile/blob/master/libqtile/notify.py) appears to provide a org.freedesktop.Notifications service, which is only used by the notify widget. It's not clear to me how I can use the service in other ways, or if I can at all. It would be cool if it could show `libqtile.window.Internal` windows or similar to the bar (i.e. unmanaged) as little notification windows. Then notify.py could have a small function to pass the service notifications that could be used in scripting, widgets, error messages and potentially many other places in qtile. What do you think?

**Comment by ramnes** [comment]
Sorry but I'm not sure to understand what you want to achieve. What's the concrete use case?

**Comment by m-col** [comment]
What is the purpose of `libqtile.notify`? I can't find any info in the docs. It is used only by the notify widget, but I figured by it being implemented outside of the widget that it was written with other uses in mind.

My idea is to glue together existing code from `libqtile.notify` and `libqtile.window` to create a graphical notifications service to show pop-up notifications much like dunst or similar.

**Comment by m-col** [comment]
I played around with this and created a 95% fully functioning graphical notification service. The code for it is [here](https://github.com/m-col/qtools/blob/master/qtools/notification/notification.py). It uses pretty much only libqtile to handle the windows. I've been using it instead of dunst for about a month and its worked really well. Here is a screenshot:

![screenshot](https://user-images.githubusercontent.com/44536113/77566907-4513f980-6ebe-11ea-9e29-12a4dac31e46.png)

For the notification service I did have to slightly [extend libqtile.notifier](https://github.com/m-col/qtile/commit/7ef41cc2123193180118a97c6b0d28a63eb8379b) so it would support all fields that notifications might supply.

It uses a base class called `Popup` ([here](https://github.com/m-col/qtools/blob/master/qtools/__init__.py#L80)), which generalises simple qtile 'internal' windows and could be useful for things like widget tooltips ([discussed here](https://github.com/qtile/qtile/issues/1441)), or maybe even to inform the user of things like config errors. If there would be interest in having this within libqtile so it can be improved and used for other windows let me know and I can create a PR, but understandable if it is outside of scope.

**Comment by pawamoy** [comment]
@m-col this is great! Though I'm not sure to understand: your screenshot shows the use of `notify-send`?

I don't get it: I already had unmanaged notification windows working before (see #1449), using notify-send, and now it doesn't work anymore (no popup appearing). Am I missing software?

**Comment by m-col** [comment]
Sorry if I wasn't clear, or if I have misunderstood you. The novel thing being done in the screenshot is not the notification client (notify-send). It is the notification server itself. The two notification windows drawn in the top corner have been created using libqtile-based code, not an external notification server such as dunst. There is a notification daemon implementation defined in `libqtile.notify`, and my plugin uses this to listen to clients such as notify-send and then draws windows with their content.

**Comment by pawamoy** [comment]
OK, I clearly didn't have the necessary knowledge to understand, thanks for your explanation :slightly_smiling_face:

**Comment by m-col** [comment]
Also if you test it and don't see the windows, you may need this commit https://github.com/m-col/qtile/commit/7ef41cc2123193180118a97c6b0d28a63eb8379b

**Comment by m-col** [comment]
re: #1449. @tych0 I just want to confirm you'd like me to submit this. If so it would come in 3 parts:

1. This commit: https://github.com/m-col/qtile/commit/7ef41cc2123193180118a97c6b0d28a63eb8379b
2. The Popup class defined [here](https://github.com/m-col/qtools/blob/master/qtools/__init__.py#L109). This would do well for the config failure notification and widget tooltips.
3. The graphical notification service defined [here](https://github.com/m-col/qtools/blob/master/qtools/notification/notification.py).

Note the TODOs in the docstring of the service. Most of these are wants, not needs, that can be added as and when they are wanted. The exception is the one that says 'overflow': I haven't yet implemented any behaviour for when there is too much text to draw on the window. Currently the text is just truncated at the edge of the window.

Thoughts? Should I submit the commit in point 1, and the Popup windows into libqtile.window, then open a third PR for discussion of where and how to integrate the notification server if that is also wanted?

**Comment by ramnes** [comment]
Sorry for not coming back to this sooner. It sounds like a very nice addition, please open the pull request(s)!

**Comment by m-col** [comment]
No worries at all - will do soon :)

**Comment by m-col** [comment]
I'm actually wondering if this wouldn't just be feature creep. I think the Popup class would be really good to have within libqtile so that it can be used for other purposes, but if that is merged then the graphical notifications work fine without any further changes to libqtile, so it would work fine as a standalone plugin. I'm happy to maintain it as I have been doing in my own plugins repo. I'm just wary that if it isn't used then it's just me adding my bloat to libqtile!

**Comment by m-col** [comment]
I'm going to close this on the basis of feature creep as per my previous comment. I'll add it to the examples repo and then link it here in case anyone else wants to use it.

**Comment by ramnes** [comment]
@m-col As discussed on IRC, feels bad to stop on such a good way. :)

**Comment by m-col** [comment]
If anybody is interested in using this or improving it, the code is available at https://github.com/m-col/qtile-config/blob/master/notification.py
